Clean architecture first attempt

exRomeo opened the pull request.

exRomeo committed:
added city, country and Weather data classes

exRomeo committed:
added image assets and working on Weather Card and SearchBar Components

exRomeo committed:
added application theme colors in the corresponding files, working on search bar

exRomeo committed:
searchbar behaviour when receives results it expands with the results list
also added the searchbar Shape in application theme files
added a function to return dummy data inside countries class just for previews

exRomeo committed:
enhanced searchbar's drop down menu changed countries api to get more results and added an interceptor to retry failed http requests

exRomeo committed:
Added ui state sealed class with cases for idle, failure, success, no network, and loading to be used to reflect the current application state on the ui also added connection and error handling utilities and added loading screen and message screen to display different messages to the user such as there is no connection or an error occurred

exRomeo committed:
removed logs and fixed a bug where state would update immediately to success when the app launches and cleaned up extra unneeded parameters

exRomeo committed:
Enhanced weather results speed with asynchronous search tasks,
also centered Text in messages screen

exRomeo committed:
Added SlantShape as a composable function

exRomeo committed:
Updated SlantShape.kt,
enhanced the layout of WeatherCard

exRomeo committed:
updated SlantShape.kt
and used it in WeatherCard

exRomeo committed:
Restructured the code to prepare for MVI Architecture and created a state holder for SearchBar still under experiment

exRomeo committed:
Handled search states Idle, Loading Error and Success and represented each on the ui also added ui events to handle presenting all error messages to the user and show them on a snackbar
used UIAction to handle all actions made from ui the screen in the view model
todo: make a SnackbarWithAction to give the user the ability to retry failed actions

exRomeo committed:
added duration "long" to the snackbar and refactored some properties in home view model

exRomeo committed:
Applied Dagger 2 Dependency Injection

exRomeo committed:
Applied Dagger Hilt Dependency Injection

exRomeo committed:
updated search bar to display the current country's name in its place holder

exRomeo committed:
first attempt at Clean Architecture

- moved repository interface to domain package
- added usecases to adhere to separation of concern and separate use cases from repository class

exRomeo committed:
Edited an Exception message to better explain the reason for the Exception

removed unnecessary state flow from HomeViewModel

exRomeo committed:
Created Domain Models for country and weather separate from network DTOs and created mappers for them also replaced the DTOs with the Domain models in Domain and Presentation Layers

removed network connectivity check as its redundant
retrofit throws IOException when there is no internet connection

in SearchBarState changed all vars to vals

exRomeo committed:
Replaced Custom class Resource with Kotlin's Result Class

exRomeo committed:
Added navigation system and navigation drawer

exRomeo committed:
updated the app's theme so the background gradient continues behind the status bar

gemy411 commented on app/build.gradle:
```diff
@@ -29,11 +34,11 @@ android {
     compileOptions {
-        sourceCompatibility JavaVersion.VERSION_1_8
-        targetCompatibility JavaVersion.VERSION_1_8
+        sourceCompatibility JavaVersion.VERSION_17
```
Is there a reason for JDK level increase?

exRomeo replied:
latest compose version needed a more recent compose compiler which needed a more recent Kotlin version which asked for a more recent jdk  so I just kept choosing the latest versions to get over the "not compatible versions" warnings and errors

short version: a series of incompatible versions led me to choose the latest jdk

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/data/apiservice/apininja/ApiNinja.kt:
```diff
@@ -0,0 +1,33 @@
+import retrofit2.http.Header
+import retrofit2.http.Query
+
+interface ApiNinja {
```
It is preferred to have a data flow for each data type.

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/data/mappers/weather/WeatherMappers.kt:
```diff
@@ -0,0 +1,21 @@
+import com.trianglz.weatherapp.data.models.weather.WeatherDto
+import com.trianglz.weatherapp.domain.models.weather.Weather
+
+fun WeatherDto.toWeather(city: CityDto): Weather =
```
In our projects, domain models have the name domain model. Like WeatherDomainModel.

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/data/mappers/weather/WeatherMappers.kt:
```diff
@@ -0,0 +1,21 @@
+import com.trianglz.weatherapp.data.models.weather.WeatherDto
+import com.trianglz.weatherapp.domain.models.weather.Weather
+
+fun WeatherDto.toWeather(city: CityDto): Weather =
```
And retrofit return model will be WeatherDataModel

gemy411 replied:
It's a convention not right or wrong.

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/data/remotesource/IRemoteDataSource.kt:
```diff
@@ -0,0 +1,27 @@
+import com.trianglz.weatherapp.data.models.country.CountryDto
+import com.trianglz.weatherapp.data.models.weather.WeatherDto
+
+interface IRemoteDataSource {
```
Our convention also, is to have the interface named like CitiesRemoteSource. and it's implementer will be CitiesRemoteSourceImpl

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/data/remotesource/IRemoteDataSource.kt:
```diff
@@ -0,0 +1,27 @@
+        limit: Int
+    ): List<CityDto>
+
+    suspend fun getWeather(
```
Don't forget to separate data flow for each type: cities, countries and weather.

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/di/modules/ApiModule.kt:
```diff
@@ -0,0 +1,55 @@
+
+@Module
+@InstallIn(SingletonComponent::class)
+class ApiModule {
```
A DI module will be provided for each data type: cities, countries and weather. To achieve separation.

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/di/modules/ApiModule.kt:
```diff
@@ -0,0 +1,55 @@
+    @Provides
+    @Named("ApiKey")
+    @Singleton
+    fun provideApiKey(): String {
```
Api key is a top level variable shouldn't be included in the minor modules. I'm thinking something application level module.

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/di/modules/UseCaseModule.kt:
```diff
@@ -0,0 +1,24 @@
+
+    @Binds
+    @Singleton
+    abstract fun bindCountrySearchUseCase(countrySearchUseCase: CountrySearchUseCase): ICountrySearchUseCase
```
On the same scheme as above. let's say we have the cities module: 
Our DI module will handle injection of the use case it's repos and it's remote sources.

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/di/modules/WeatherAppModule.kt:
```diff
@@ -0,0 +1,39 @@
+
+    @Binds
+    @Singleton
+    abstract fun bindsExceptionHandler(exceptionHandler: ExceptionHandler): IExceptionHandler
```
Please clarify to me why the exception handler is injected.

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/domain/models/weather/Weather.kt:
```diff
@@ -0,0 +1,37 @@
+        else -> "Stay indoors."
+    }
+
+    fun getWeatherIcon(): Int {
```
Weather icon is UI related property. It is preferred to have a separate model for the UI related aspects.

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/domain/usecases/countrysearch/CountrySearchUseCase.kt:
```diff
@@ -0,0 +1,29 @@
+) : ICountrySearchUseCase {
+    override suspend fun getCountries(
+        countryName: String
+    ): Result<List<Country>> = coroutineScope {
```
Kotlin already provides a function for handling such matter. runCatching { }. can be used to create a result object out of the box.

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/domain/usecases/countrysearch/ICountrySearchUseCase.kt:
```diff
@@ -0,0 +1,9 @@
+
+import com.trianglz.weatherapp.domain.models.country.Country
+
+interface ICountrySearchUseCase {
```
Please clarify the need to have an interface for a use case.

exRomeo replied:
to make testing the view model easier

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/domain/usecases/getweatherdata/WeatherDataUseCase.kt:
```diff
@@ -0,0 +1,41 @@
+    override suspend fun getWeatherData(
+        countryCode: String,
+        limit: Int
+    ): Result<List<Weather>> = coroutineScope {
```
Very neat. But, what I'd do is have a use case for fetching the weather data for each city. And from the view model I'd loop and fetch multiple weather datas for each city.

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/domain/utils/exceptionhandler/ExceptionHandler.kt:
```diff
@@ -0,0 +1,24 @@
+        else -> "An error occurred. Please try again later."
+    }
+
+    private fun handleHttpException(exception: HttpException): String = when (exception.code()) {
```
Great approach!!

gemy411 reviewed (commented):
Amazing work so far! Keep it up!

exRomeo committed:
Mass Refactoring
all data package modes renamed to have 'DataModel' in its name instead of 'Dto'

all Domain Models renamed to have 'DomainModel' in its name

renamed mappers according to the models new names

also renamed all interface implementations to have 'Impl' in its name and removed the 'I' from the interfaces names

also updated Compose Dependencies which required an sdkTarget upgrade which also required gradle upgrade

exRomeo committed:
Refactored modules to be feature specific

exRomeo committed:
saving

exRomeo committed:
Separated Country,City and Weather Flows with separate use cases

exRomeo committed:
made all use cases return result instance and handled success and failure cases in the view model

exRomeo committed:
Refactored the functions responsible for getting cities and weather data to handle successful and failure cases in a better way

exRomeo committed:
added documentation to the code

also removed the utility manager class -> not used any more

exRomeo committed:
moved ExceptionHandler.kt to Presentation layer
renamed it to ExceptionResolver
and modified it to return localized string from string resource instead of a static English string
also added a provide function in the app module to provide an instance of Resources class to use it in the ExceptionResolver

exRomeo committed:
added notifications support and a simple work manager to show a notification each 15 minutes

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/data/apiservice/apininja/ApiNinja.kt:
```diff
@@ -0,0 +1,26 @@
+        @Query("country") countryCode: String,
+        @Query("limit") limit: Int
+    ): List<CityDataModel>
+
```
Still having two function in the API ninja interface. On a larger scale this won't support extension and furhter mutations.

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/presentation/exceptionresolver/ExceptionResolverImpl.kt:
```diff
@@ -0,0 +1,43 @@
+import java.net.SocketTimeoutException
+import javax.inject.Inject
+
+class ExceptionResolverImpl @Inject constructor(private val resourceManager: Resources) :
```
The pinnacle of human engineering XD 
Love it

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/presentation/exceptionresolver/ExceptionResolverImpl.kt:
```diff
@@ -0,0 +1,43 @@
+import javax.inject.Inject
+
+class ExceptionResolverImpl @Inject constructor(private val resourceManager: Resources) :
+    ExceptionResolver {
```
If you want to achieve better separation. Here is a tip: 
It's a good decision to have this in the presentation layer. 
But talking about HTTP errors in UI layers isn't optimum. 
The ideal flow of thought would be that, the domain shoots for the data layer, while data layer errors are mapped to domain layer errors, and those domain layer errors are mapped to UI String in the UI layer. 
It's not necessary but it's how the chain of though in the clean arch works. Still, this  code works perfectly and still very modular and maintainable. So, good job!!

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/domain/repository/city/CitiesRepository.kt:
```diff
@@ -0,0 +1,11 @@
+    suspend fun getCities(
+        countryCode: String,
+        limit: Int
+    ): List<CityDataModel>
```
This repository declaration is a domain specidifc object. So, referencing data models isn't recommended.

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/di/modules/city/CityModule.kt:
```diff
@@ -0,0 +1,36 @@
+
+@Module
+@InstallIn(SingletonComponent::class)
+abstract class CityModule {
```
perfect execution

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/domain/usecases/fetchcities/FetchCitiesUseCase.kt:
```diff
@@ -0,0 +1,16 @@
+    suspend fun getCities(
+        countryCode: String,
+        limit: Int
+    ): Result<List<CityDataModel>>
```
Data models are data layer specific. 
Here we use domain models.

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/presentation/ui/home/HomeViewModel.kt:
```diff
@@ -0,0 +1,216 @@
+    ) = coroutineScope {
+        val weatherResult =
+            cityList.map { city ->
+                async {
```
Very clean! Having async the return of map. Love it.

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/presentation/extensions/ModifierExtensions.kt:
```diff
@@ -0,0 +1,17 @@
+ * @param ratio is the percentage of screen's height the component's maximum height will be
+ * */
+
+fun Modifier.maxHeightScreenRatio(@FloatRange(0.2, 1.0) ratio: Float) = composed {
```
Good convention.

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/presentation/ui/components/WeatherCard.kt:
```diff
@@ -0,0 +1,237 @@
+import com.trianglz.weatherapp.presentation.ui.theme.lavender
+
+@Composable
+fun WeatherCard(
```
Why does the weather card do not have a data class for it's state like the search bar?

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/presentation/searchbarstate/SearchBarState.kt:
```diff
@@ -0,0 +1,41 @@
+    placeHolder: String = "",
+    noResultMessage: String = "",
+    status: SearchBarStatus = SearchBarStatus.Idle,
+    result: List<CountryDomainModel> = emptyList(),
```
The search bar should be decoupled from the list of results (Countries)

gemy411 reviewed (changes requested):
Keep up the great work! This app is almost production ready!

exRomeo committed:
separated weather and city Api interfaces
created a model for the weather card
decoupled the search result from the search bar state
